### PR TITLE
PERF: Add __contains__ to CategoricalIndex

### PR DESCRIPTION
**Pull request by topper-123**
- [x] progress towards #20395
- [x] xref #21022
- [ ] tests added / passed
- [x] benchmark added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Currently, membership checks in ``CategoricalIndex`` is very slow as explained in #21022. This PR fixes the issue for ``CategoricalIndex``, while #21022 contains the fix for ``Categorical``. The difference between the two cases is the use of ``_engine`` for ``CategoricalIndex``, which makes this even faster than the ``Catagorical`` solution in #21022.

Tests exist already and can be found in ``tests/indexes/test_category.py::TestCategoricalIndex::test_contains``.

ASV:

```
      before           after         ratio
     [0c65c57a]       [986779ab]
-      2.49±0.2ms       3.26±0.2μs     0.00  categoricals.Contains.time_contains

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

### CODE REVIEW

File: pandas/core/indexes/category.py
Comment (gfyoung) [inline pandas/core/indexes/category.py]: This might be a really silly question, but what does this line do?

File: pandas/core/indexes/category.py
Comment (gfyoung) [inline pandas/core/indexes/category.py]: Remind me: why do we need to check membership in `self.values` anymore?

File: pandas/core/indexes/category.py
Comment (topper-123) [reply pandas/core/indexes/category.py]: It just ensures that mutables are not passes into the function....

I think it should probably be removed, but that line is in various other places as well, so maybe a seperate PR, that goes through all similar cases? Or I can just remove it.

Comment (gfyoung) [reply pandas/core/indexes/category.py]: Well, you removed it in another part of the diff, hence why I'm asking.  Let's investigate for another time then, in which case I would put back the other one that you deleted.

File: pandas/core/indexes/category.py
Comment (topper-123) [reply pandas/core/indexes/category.py]: For indices, their indexing engine (i.e. ``._engine``) has a ``__contains__`` method which does the same thingbut  is faster.

Comment (gfyoung) [reply pandas/core/indexes/category.py]: Awesome, thanks for clarifying!

**Review by gfyoung — APPROVED** [review]
Nice!

cc @jreback

File: pandas/core/indexes/category.py
Comment (jreback) [inline pandas/core/indexes/category.py]: can you put a blank line between things, e.g.
```
if isna(...):
     ....

try:
   ....
except:
  ....

if is_scalar(...):
   ...

# no else needed here
return ...
```

Comment (jreback) [reply pandas/core/indexes/category.py]: also can you put comments before each case (not everythin needs a comment), but i find this hard to grok in its current form.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
I've updated the PR.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @topper-123

**Comment by jorisvandenbossche** [comment]
This is changing the implementation quite substantially, so let's move this to 0.24.0.txt?

**Comment by jorisvandenbossche** [comment]
Any comments on my comment above about keeping this for 0.24.0 ?

**Comment by topper-123** [comment]
Quility-wise this is ok to go into 23.2 IMO, the PRs are really not that complex, IMO, it's much faster and it doesn't change any APIs.

Also, my main motivation for writing this was speeding up slicing dataframes with a CategoricalIndex (see #20395), which previously was *very* slow (still is slow, but better than before, and now faster than fancy indexing, at least). I think a lot of people will appreciate this speedup.

**Comment by jreback** [comment]
we tagged for 0.23.2 (and note is there)
it would be slightly tricky to change as there is another related change - both in 0.23.2 or 0.24
either way is fine

**Comment by jorisvandenbossche** [comment]
The main reason that I am raising this is that `__contains__` checking has quite some implications (which is of course also the reason you speeded it up!), and I think it is rather easy to miss a small exotic corner case where the new implementation might differ. 
To be clear, @topper-123 , I am not questioning the quality of this PR! I just know from experience, also for us, that it is easy to miss an unintended API change (which we might not even decide to fix if it is debateble behaviour, but then that is still better to be left as 0.24.0). Since this is only about performance improvement (and not a regression), I would play on safe and give it more time in 0.24.0.


>  it would be slightly tricky to change as there is another related change - both in 0.23.2 or 0.24

Yep, it would be both in 0.23.2 or both in 0.24.0 (the other I actually tagged first as 0.24.0, but that was changed before merging)

**Comment by topper-123** [comment]
Ok, your call, I won't object.

**Pull request merged**
